RecoTracker/LSTCore/src/alpaka/Event.dev.cc

```diff
@@ -0,0 +1,1824 @@
+    delete quintupletsInGPU;
+    delete quintupletsBuffers;
+    quintupletsInGPU = nullptr;
+  }
+  if (trackCandidatesInGPU) {
+    delete trackCandidatesInGPU;
+    delete trackCandidatesBuffers;
+    trackCandidatesInGPU = nullptr;
+  }
+  if (pixelTripletsInGPU) {
+    delete pixelTripletsInGPU;
+    delete pixelTripletsBuffers;
+    pixelTripletsInGPU = nullptr;
+  }
+  if (pixelQuintupletsInGPU) {
+    delete pixelQuintupletsInGPU;
+    delete pixelQuintupletsBuffers;
+    pixelQuintupletsInGPU = nullptr;
+  }
+
+  if (hitsInCPU != nullptr) {
+    delete hitsInCPU;
+    hitsInCPU = nullptr;
+  }
+  if (rangesInCPU != nullptr) {
+    delete rangesInCPU;
+    rangesInCPU = nullptr;
+  }
+  if (mdsInCPU != nullptr) {
+    delete mdsInCPU;
+    mdsInCPU = nullptr;
+  }
+  if (segmentsInCPU != nullptr) {
+    delete segmentsInCPU;
+    segmentsInCPU = nullptr;
+  }
+  if (tripletsInCPU != nullptr) {
+    delete tripletsInCPU;
+    tripletsInCPU = nullptr;
+  }
+  if (quintupletsInCPU != nullptr) {
+    delete quintupletsInCPU;
+    quintupletsInCPU = nullptr;
+  }
+  if (pixelTripletsInCPU != nullptr) {
+    delete pixelTripletsInCPU;
+    pixelTripletsInCPU = nullptr;
+  }
+  if (pixelQuintupletsInCPU != nullptr) {
+    delete pixelQuintupletsInCPU;
+    pixelQuintupletsInCPU = nullptr;
+  }
+  if (trackCandidatesInCPU != nullptr) {
+    delete trackCandidatesInCPU;
+    trackCandidatesInCPU = nullptr;
+  }
+  if (modulesInCPU != nullptr) {
+    delete modulesInCPU;
+    modulesInCPU = nullptr;
+  }
+}
+
+void SDL::Event<SDL::Acc>::addHitToEvent(std::vector<float> x,
+                                         std::vector<float> y,
+                                         std::vector<float> z,
+                                         std::vector<unsigned int> detId,
+                                         std::vector<unsigned int> idxInNtuple) {
+  // Use the actual number of hits instead of a max.
+  unsigned int nHits = x.size();
+
+  // Initialize space on device/host for next event.
+  if (hitsInGPU == nullptr) {
+    hitsInGPU = new SDL::hits();
+    hitsBuffers = new SDL::hitsBuffer<Dev>(nModules_, nHits, devAcc, queue);
+    hitsInGPU->setData(*hitsBuffers);
+  }
+
+  if (rangesInGPU == nullptr) {
+    rangesInGPU = new SDL::objectRanges();
+    rangesBuffers = new SDL::objectRangesBuffer<Dev>(nModules_, nLowerModules_, devAcc, queue);
+    rangesInGPU->setData(*rangesBuffers);
+  }
+
+  // Need a view here before transferring to the device.
+  auto nHits_view = alpaka::createView(devHost, &nHits, (Idx)1u);
+
+  // Copy the host arrays to the GPU.
+  alpaka::memcpy(queue, hitsBuffers->xs_buf, x, nHits);
+  alpaka::memcpy(queue, hitsBuffers->ys_buf, y, nHits);
+  alpaka::memcpy(queue, hitsBuffers->zs_buf, z, nHits);
+  alpaka::memcpy(queue, hitsBuffers->detid_buf, detId, nHits);
+  alpaka::memcpy(queue, hitsBuffers->idxs_buf, idxInNtuple, nHits);
+  alpaka::memcpy(queue, hitsBuffers->nHits_buf, nHits_view);
+  alpaka::wait(queue);
```

Comment:
The inputs are on the stack (they are copied during the function call), so they will go out of scope when the function returns.

Reply:
A somewhat better approach could be to
  - create a host buffer in pinned memory using the same `queue`;
  - `std::memcpy` synchronously the data from the `std::vector` to the host buffer;
  - `alpaka::memcpy` asynchronously the host buffer to the device buffer.

Of course a _much better_ approach would be to fill a host SoA directly in `LSTPhase2OTHitsInputProducer`, copy it to device, and avoid the 6 intermediate copies of the `std::vector`s.

Reply:
> A somewhat better approach could be to
> 
>     * create a host buffer in pinned memory using the same `queue`;
> 
>     * `std::memcpy` synchronously the data from the `std::vector` to the host buffer;
> 
>     * `alpaka::memcpy` asynchronously the host buffer to the device buffer.

somewhat simplified, what we have now is 
```C++
void lst::Event::addHitToEvent(std::vector<unsigned int> const& detId){
  alpaka::memcpy(queue, hitsBuffers->detid_buf, detId, nHits);
  alpaka::wait(queue);
}
void LST::run(TQueue, Ph2_vectors_edm, seed_vectors_edm) {
// fills a data member in_hitId_ , combined from Ph2 and Seeds
  auto event = lst::Event<Acc3D>(queue, deviceESData);
  event.addHitToEvent(in_hitId_);
...
  event.createMiniDoublets(); // uses hits
}
void LSTProducer::acquire(device::Event const& ev, device::EventSetup const& setup)  {
  lst_.run(ev.queue(), ...
```

are we actually missing creation/copy of the `detId` in the pinned memory or does the `wait` serve as an effective lock allowing the `alpaka::memcpy` to complete correctly?

If a pinned memory buffer is created, where should it be done to have a correct lifetime? Should it be a member of `lst::Event` or be defined in `LST::run` scope?
(if I understand correctly, it can't be on stack inside `lst::Event::addHitToEvent`, at least not without a `wait`)

Reply:
> are we actually missing creation/copy of the `detId` in the pinned memory or does the `wait` serve as an effective lock allowing the `alpaka::memcpy` to complete correctly?

If I understood your question correctly, at least towards correctness the `wait()` guarantees the `alpaka::memcpy()` will finish before the `detId` is destructed (even if `detId` would be a temporary).

> If a pinned memory buffer is created, where should it be done to have a correct lifetime? Should it be a member of `lst::Event` or be defined in `LST::run` scope?
> (if I understand correctly, it can't be on stack inside `lst::Event::addHitToEvent`, at least not without a `wait`)

When created with `cms::alpakatools::make_host_buffer()`, the buffer actually can be in the stack of `lst::Event::addHitToEvent()`. By default the caching allocator will keep the actual pinned host memory buffer alive and reserved (at least) until the `alpaka::memcpy()` has finished. If the use of caching allocator is disabled, the pinned host memory buffer will itself synchronize in its destructor.

Reply:
> > are we actually missing creation/copy of the `detId` in the pinned memory or does the `wait` serve as an effective lock allowing the `alpaka::memcpy` to complete correctly?
> 
> If I understood your question correctly, at least towards correctness the `wait()` guarantees the `alpaka::memcpy()` will finish before the `detId` is destructed (even if `detId` would be a temporary).

The earlier comment from Andrea was suggesting to use a pinned memory host buffer (IIUC, a concept different from the buffer being attached to a queue).
So, my last question expanded the scope of just the lifetime of the variable/buffer in the function scope to the underlying host memory property.
Just to rephrase, is there supposed to be also a copy from `detId` (`std::vector`) to some pinned memory buffer in the code discussed here
```C++
  alpaka::memcpy(queue, hitsBuffers->detid_buf, detId, nHits);
  alpaka::wait(queue);
```
?
or is pinning essential only with asynchronous copying. ... or also if cache is used

From [a web blog](https://forums.developer.nvidia.com/t/advantages-disadvantages-of-using-pinned-memory/34422/7) I read that pinning has an extra cost and makes sense if the buffer is reused.

Reply:
> or is pinning essential only with asynchronous copying

Pinning is generally essential for asynchronous copies.

> From [a web blog](https://forums.developer.nvidia.com/t/advantages-disadvantages-of-using-pinned-memory/34422/7) I read that pinning has an extra cost and makes sense if the buffer is reused.

Correct, pinning has a cost.

> ... or also if cache is used

The cost of the pinning is the reason why we use caching allocator also for the pinned host memory. Another reason is that with the caching allocator you don't have to worry that much about the life time of the pinned host memory buffer, because the buffer can be "freed" (i.e. given to the allocator to manage) immediately after the host-to-device copy has been issued.

----

> Just to rephrase, is there supposed to be also a copy from `detId` (`std::vector`) to some pinned memory buffer in the code discussed here
> 
> ```c++
>   alpaka::memcpy(queue, hitsBuffers->detid_buf, detId, nHits);
>   alpaka::wait(queue);
> ```
> ?

In order to make this snippet asynchronous, yes, a copy to a pinned memory buffer is needed. It could be along
```cpp
  // is detId.size() == nHits ?
  auto detid_buf_h = cms::alpakatools::make_host_buffer<unsigned int[]>(queue, detId.size());
  std::copy(detId.begin(), detId.end(), detid_buf_h.data());
  alpaka::memcpy(queue, hitsBuffers->detid_buf, detid_buf_h);
```